# Python assignment solution

1. Calculate Area of a Circle

In [4]:
import math

def circle_area(radius):
    return math.pi * radius ** 2

In [5]:
circle_area(10)

314.1592653589793

2. Say Hello to user, Expected, ‘Hello name’

In [6]:
def hello(name):
    print(f'Hello {name}')

In [7]:
hello("Diponkar")

Hello Diponkar


3. Calculate Volume of Sphere

In [8]:
import math
def sphere_volume(radius):
    return (4/3) * math.pi * radius ** 3

In [9]:
sphere_volume(4)

268.082573106329

4. Copy String in n times

In [14]:
def copy_string(string,n):
    for i in range(n):
        print(str(i+1) + '. ' + string)

In [15]:
copy_string('String', 5)

1. String
2. String
3. String
4. String
5. String


5. Calculate Area of a Triangle

In [27]:
#Sol_1
def triangle_area(base, height):
    """ This function calculate the triangle area based on the base and the height."""
    return (1/2) * base * height

In [28]:
triangle_area(20, 10)

100.0

In [29]:
#Sol_2
import math
def triangle_area(edge1, edge2, edge3):
    """ This function calculate the triangle area based on three edge."""
    if edge1 + edge2 < edge3 or edge2 + edge3 < edge1 or edge1 + edge3 < edge2:
        return " This is not a triangle. A triangle must have the summation of two edge is bigger than third edge."
    semi_perimeter = (edge1 + edge2 + edge3) / 2
    triangle_area = math.sqrt(semi_perimeter * (semi_perimeter - edge1) * (semi_perimeter - edge2) *(semi_perimeter - edge3))
    return triangle_area

In [30]:
triangle_area(12,34,5)

' This is not a triangle. A triangle must have the summation of two edge is bigger than third edge.'

In [31]:
triangle_area(10,12,13)

56.99506557588999

6. Calculate Interest

In [32]:
def interest_calculator(principal, int_rate, time):
    """This function calculate the interest.
    Formula for interest is:
    Principal * Interest rate * Time 
    Here time is usually given in year"""
    return principal * int_rate * time

In [33]:
interest_calculator(2000, 0.05, 5)

500.0

7. calculate Euclidean Distance

In [38]:
import math
def euclidean_distance(x1,y1,x2, y2):
    """Euclidean distance is distance between to objects.
    Two objects point (x1, y1), (x2, y2).
    The formula for mesuring the distance between these two points is:
    sqrt((x2-x1)**2 + (y2 - y1) **2)"""

    return math.sqrt((x2 - x1)**2 + (y2 - y1)**2)

In [40]:
euclidean_distance(4,4,10,10)

8.48528137423857

8. Feet to centimeter Converter

In [1]:
def feet_to_centimeter(feet_distince):
    """This function takes input as fee distance and return the coverted distance in centimeter"""
    return feet_distince * 30.48

In [3]:
feet_to_centimeter(4)

121.92

9. BMI Calculator

In [4]:
def bmi_calculator(weight, height):
    """ BMI: Body mass index
    formula: weight(kg)/ meter**2"""
    return weight / (height ** 2)

In [5]:
bmi_calculator(90,1.59)

35.59985760056959

In [6]:
bmi_calculator(64,1.67)

22.948115744558788

10. Binary to Decimal

In [24]:
def bin_to_dec(bin):
    """ This function convert the binary value to decimal"""
    sbin = str(bin)
    re_sbin=  sbin[:: -1]
    dec = 0
    for i in range(len(re_sbin)):
        dec += int(re_sbin[i])* 2 ** i
    return dec


In [25]:
bin_to_dec(101111)

47

11. receive two integral numbers in string form and compute their sum and then print it in console.

In [26]:
def summation(num1, num2):
    """ Function to compute summation of two integral string"""
    print('Summation: ', str(int(num1) + int(num2)))

In [27]:
summation('10','20')

Summation:  30


12. bitwise operators for any two variable all the bitwise operator that are there

#### Bitwise operator and meaning
- & --> Bitwise AND
- | --> Bitwise OR
- ^ --> Bitwise XOR
- ~ --> Bitwise NOT
- << --> Bitwise Left shift
- >> --> Bitwise right shift

In [34]:
a = 10
b = 11
c = a & b
c

10

In [36]:
c = a|b
c

11

In [37]:
c = a ^ b
c

1

In [39]:
c = ~ b
c

-12

In [41]:
c  = a << b
c

20480

In [42]:
c = a >> b
c

0

13. shift operators

In [43]:
# right shift operator
a = 5
b = 5
c = a>>b
c

0

In [44]:
# left shift operator

c = a << b
c

160

14. str1=" this is a sample daTaa to test for the task" Check if ‘is’ present in str1

In [29]:
str1 = " this is a sample daTaa to test for the task"
def is_exist(str):
    number_of_is = str.find('is')
    return number_of_is > 0

In [30]:
is_exist(str1)

True

15.  Check if ‘was’ not present in str1

In [31]:
def was_exist(str):
    number_of_was = str.find('was')
    return number_of_was > 0

In [32]:
was_exist(str1)

False

16. Print str1 in uppercase

In [45]:
def string_uppercase(str):
    return str.upper()


In [46]:
string_uppercase("data Science")

'DATA SCIENCE'

17.  Print str1 as title

In [47]:
def string_title(str):
    return str.title()

In [49]:
string_title('machine learning and data science')

'Machine Learning And Data Science'

18. Get a new string from str1 ‘DatTaa’ as ‘Data’

In [50]:
str1 = " this is a sample daTaa to test for the task"
new_string = str1.replace('daTaa','Data')
print(new_string)

 this is a sample Data to test for the task


19. Print a new string with extra "a" removed form previous task

In [53]:
str1 = " this is a sample daTaa to test for the task"
print(str1.replace(' a ', ' '))

 this is sample daTaa to test for the task


20. Display ‘sample’ word from ‘ this is another sample string ‘

In [55]:
str_20 = 'this is another sample string'
str_sample_loc = str_20.find('sample')
str_sample = str_20[str_sample_loc: str_sample_loc+6]
print(str_sample)

sample


21. Display date, month and year form 21-oct-2020

In [62]:
date = '21-oct-2020'
date_list = date.split('-')
print(f'Day: {date_list[0]},  Month: {date_list[1]}, and  Year: {date_list[2]}')

Day: 21,  Month: oct, and  Year: 2020


22. arr = “10,20,30,40,50,60,70”, get “50,60”

In [64]:
arr = "10,20,30,40,50,60,70"
loc_target_str = arr.find('50,60')
target_str = arr[loc_target_str:loc_target_str+5]
target_str

'50,60'

23. using negative index get "50 ,60"

In [67]:
arr = "10,20,30,40,50,60,70"
target_str = arr[-8:-3]
target_str

'50,60'

24. From credit_str get ‘8888’
credit_str = "xxxx----xxxx-8888------xxxx"

In [69]:
credit_str = "xxxx----xxxx-8888------xxxx"
loc_target_str = credit_str.find('8888')
target_str = credit_str[loc_target_str:loc_target_str+4]
target_str


'8888'

25. From credit_str get ‘0’
credit_str = "1234-5678-9878-0434"

In [70]:
credit_str = "1234-5678-9878-0434"
target_str = credit_str[-4:-3]
target_str

'0'

26. convert a integer into a string and print it in console.

In [72]:
int_number =int(input('Enter a integer number: '))
str_number = str(int_number)
print('String Number: ', str_number)

String Number:  45


27. take two strings as input and concatenate them and then print it

In [73]:
str1 = str(input("Enter first string: "))
str2 = str(input("Enter first string: "))
concateneted_str = str1+str2
print('Concatenated String: ', concateneted_str)

Concatenated String:  Machine Lerningand Data Science


28. convert the given string into list
given_string = " Apples is differnt from Apple Inc"

In [76]:
given_string = " Apples is differnt from Apple Inc"

def string_to_list(str):
    new_list = list(str.split())
    return new_list

In [77]:
string_to_list(given_string)

['Apples', 'is', 'differnt', 'from', 'Apple', 'Inc']

29. print type of a and get the 1 element
a = {(1,2):1,(2,3):2}

In [82]:
a = {(1,2):1,(2,3):2}
print("Type of variable a: ", type(a))
print(f'Dictionary variables a 1 element: {a[1,2]}')

Type of variable a:  <class 'dict'>
Dictionary variables a 1 element: 1


30. write a program to get  list as 4,3,2 from the list below
a=1,2,3,4,5

In [84]:
a=[1,2,3,4,5]
new_list= a[1:4]
out_list = new_list[::-1]
out_list

[4, 3, 2]

31. write a program to get odd number only using a slicing
a=1,2,3,4,5,6,7,8,9

In [86]:
a=[1,2,3,4,5,6,7,8,9]
odd_num_list = a[0:8:2]
odd_num_list

[1, 3, 5, 7]

32. create a tuple with (1,2) 5 times

In [87]:
new_tuple = (1,2) * 5
new_tuple

(1, 2, 1, 2, 1, 2, 1, 2, 1, 2)

33. from given list create separate lists of strings and numbers.use list comprehension
gadgets = "Mobile", "Laptop", 100, "Camera", 310.28, "Speakers", 27.00, "Television", 1000, "Laptop Case", "Camera Lens"

In [91]:
gadgets = ["Mobile", "Laptop", 100, "Camera", 310.28, "Speakers", 27.00, "Television", 1000, "Laptop Case", "Camera Lens"]

str_list = [ item for item in gadgets if type(item) == str]
num_list = [ item for item in gadgets if type(item) != str]
print(f'String List: {str_list} \n Number List: {num_list}')


String List: ['Mobile', 'Laptop', 'Camera', 'Speakers', 'Television', 'Laptop Case', 'Camera Lens'] 
 Number List: [100, 310.28, 27.0, 1000]


34. An array is given from the array find the sum of diagonal using list comprehension
a_2d=[1, 2, 3, 4, 5, 6, 7, 8, 9]

In [96]:
a_2d=[[1, 2, 3], [4, 5, 6], [7, 8, 9]]
sum_diagonal = 0
for i in range(len(a_2d)):
    for j in range(len(a_2d[i])):
        if i == j:
            sum_diagonal += int(a_2d[i][j])
print(f'Summation of diagonal elements: {sum_diagonal}')


Summation of diagonal elements: 15


35. convert number into tuple of words using comprehension
num=8976543201

In [106]:
num='8976543201'
tuple( [int(d) for d in num])


(8, 9, 7, 6, 5, 4, 3, 2, 0, 1)

36. print the numbers which are divisible by 2 but are not a multiple of 5
numbers = range(1, 10)

In [1]:
for num in range(0,10,2):
    if num%2==0 and num%5 != 0:
        print(num)



2
4
6
8


37. print the sum of tuple : use comprehension
init_tuple = [(0, 1), (1, 2), (2, 3)]

In [2]:
init_tuple = [(0, 1), (1, 2), (2, 3)]
sum_init_tuple = [sum(tup) for tup in init_tuple]
sum_init_tuple

[1, 3, 5]

38. delete the last coumn from the given 4*4 matrix
arr = [[1, 2, 3, 4],
       [4, 5, 6, 7],
       [8, 9, 10, 11],
       [12, 13, 14, 15]]

In [15]:
#Sol1
# import numpy as np
# arr = [[1, 2, 3, 4],
#        [4, 5, 6, 7],
#        [8, 9, 10, 11],
#        [12, 13, 14, 15]]
# arr = np.array(arr)       
# new_arr = arr[:,0:3]
# new_arr    
  

array([[ 1,  2,  3],
       [ 4,  5,  6],
       [ 8,  9, 10],
       [12, 13, 14]])

In [22]:
#Sol2
arr = [[1, 2, 3, 4],
       [4, 5, 6, 7],
       [8, 9, 10, 11],
       [12, 13, 14, 15]]     

In [23]:
for i in range(len(arr)):
    arr[i].pop()# the logic behind it is , pop function always pop the last item of the list
arr

[[1, 2, 3], [4, 5, 6], [8, 9, 10], [12, 13, 14]]

39. get the list of value of dict in sorted way
dict = {'c': 97, 'a': 96, 'b': 98}

In [24]:
dict = {'c': 97, 'a': 96, 'b': 98}
sorted_dict_list = sorted(dict.values())
sorted_dict_list

[96, 97, 98]

40. Creating a Dictionary with Integer Keys and print it

In [25]:
dict_int_keys = {1:'python',2:'SQL',3:'R'}
dict_int_keys

{1: 'python', 2: 'SQL', 3: 'R'}

121:Write a function which implements the Pascal's triangle use recursion:(https://www.mathsisfun.com/pascals-triangle.html#:~:text=One%20of%20the%20most%20interesting,directly%20above%20it%20added%20together.)



122: Write a recursive program to calculate the Fibonacci numbers, using Pascal's triangle.

123: write a fuction  to check if the given data type are iterable or not ?

124: Write a generator  which computes the running average.
```
    eg:[1,2,3,4,5,6]
        1-->1
        2---->1.50
        3----->2
        4----->2.50
        5---->3
        6---->3.50
 ```

125: Write a generator  which behaves like range but accepts float values.

126:Write a generator which returns  a zero or a one in every iteration. 
- The probability p for returning a 1 is defined in a variable p.
- default value of probability=0.5. In other words,1 and 0 will be returned with the same probability.

127: Write a generator cycle performing the same task.

- cycle('ABCD') --> A B C D A B C D A B C D ...

128)  create a list of list from the given detail below
```
id   product              Quantity     Price_per
1020 iphone 12          4          1200
2313 mac book pro       5          3200                
1234 drone              3          900
3212 tv                 8          500
3213 book               3          20
````


Write a Python program, which returns a list with 2-tuples. Each tuple consists of a the product name and  the total price of product . The product should be increased by 10,- € if the value of the order is smaller than 100,00 €.
Write a Python program using lambda and map.

129: Write a program which returns a list of two tuples with (order number, total amount of order).

- list of tuple:
```
[ordernumber, (article number, quantity, price per unit), ... (article number, quantity, price per unit) ]
```

use : map, reduce and lambda

In [26]:
orders = [ [1, ("5464", 4, 10), ("8274",18,15), ("9744", 9, 20)], 
           [2, ("5464", 9, 9.99), ("9744", 9, 44.95)],
           [3, ("5464", 9, 9.99), ("88112", 11, 24.99)],
           [4, ("8732", 7, 11.99), ("7733",11,18.99), ("88112", 5, 39.95)],
            [5, ("8732", 7, 11.99), ("7733",11,18.99), ("8274", 5, 39.95)],
          [6, ("8732", 7, 11.99), ("8274",11,18.99), ("9744", 5, 39.95)] 
         ]

--output```[[1, 490], [2, 494.46000000000004], [3, 364.79999999999995], [4, 492.57], [5, 492.57], [6, 492.57]]````

130: write a nested function to calulate the temp from celcius to farhen

131:Write a nested  function so that it can work for polynomials of arbitrary degree:

\begin{equation}
\sum_{k=0}^{n} a_{k} \cdot x^{k}=a_{n} \cdot x^{n}+a_{n-1} \cdot x^{n-1}+\ldots+a_{2} \cdot x^{2}+a_{1} \cdot x+a_{0}
\end{equation}

132:write a decorator so that the argument passed is. positive integer for a function factorial that you have created before

133:write a decorator to count the number of times a function has beeen called . eg
define a add function and loop it for 10 time and check